server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -115,4 +128,30 @@ public void onFailure(Exception exception) {
             );
         }
     }
+
+    private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
+        SearchSourceBuilder source = request.source();
+        if (source == null || source.knnSearch() == null) {
+            return request;
+        }
+
+        List<ScoreDoc> scoreDocs = new ArrayList<>();
+        for (ScoreDoc scoreDoc : knnResults.scoreDocs()) {
+            if (scoreDoc.shardIndex == request.shardRequestIndex()) {
+                scoreDocs.add(scoreDoc);
+            }
+        }
+        scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
+        KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
+
+        SearchSourceBuilder newSource = source.shallowCopy().knnSearch(null);
+        if (source.query() == null) {
+            newSource.query(knnQuery);
+        } else {
+            newSource.query(new BoolQueryBuilder().should(knnQuery).should(source.query()));
+        }
```

Comment:
Should we specialize the case when no global vectors belong to one shard either here or through implementing `rewrite` on `KnnDocsQueryBuilder`?

Reply:
We do implement `KnnScoreDocQuery#rewrite`, we could implement `KnnScoreDocQueryBuilder#rewrite` for completeness too. This query is never used in `can_match` so not sure it makes a difference right now 🤔

Reply:
Agreed, I think we disable caching too with `DFS_QUERY_THEN_FETCH`, so it shouldn't matter.